File: api/src/classes/Terrain.js
class Terrain {
  constructor({category, type}) {
    this.category = category
    this.type = type

    this.defenseModifier = 0
    this.isPassable = true
    this.isObstructed = false

    switch(category) {
      case 'standard':
        break
      case 'protected':
        this.defenseModifier = 1
        this.isObstructed = true
        break
      case 'exposed':
        this.defenseModifier = -1
        break
      case 'impassable':
        this.isPassable = false
        this.isObstructed = true
        break
    }
  }

  toObj(slim=false) {
    let essential = {
      category: this.category,
      type: this.type
    }
    let extra = {
      defenseModifier: this.defenseModifier,
      isPassable: this.isPassable,
      isObstructed: this.isObstructed
    }
    return slim ? essential : {...essential, ...extra}
  }
}

module.exports = Terrain

File: rails/api/lib/terrain.rb
class Terrain
  attr_reader :type, :category, :defense_modifier, :offense_modifier

  def initialize(category: nil, type: nil)
    case category&.to_sym&.downcase
    when :exposed
      @category = :exposed
      @type = type ? type.to_sym&.downcase : :road
      @defense_modifier = -1
      @offense_modifier = 0
      @passable = true
      @obstructs = false
    when :impassable
      @category = :impassable
      @type = type&.to_sym&.downcase || :mountain
      @defense_modifier = 0
      @offense_modifier = 0
      @passable = false
      @obstructs = true
    when :protected
      @category = :protected
      @type = type&.to_sym&.downcase || :forest
      @defense_modifier = 1
      @offense_modifier = 0
      @passable = true
      @obstructs = true
    else
      @category = :standard
      @type = type&.to_sym&.downcase || :plains
      @defense_modifier = 0
      @offense_modifier = 0
      @passable = true
      @obstructs = false
    end
  end

  def to_h
    {
      category: category,
      type: type,
      defense_modifier: defense_modifier,
      offense_modifier: offense_modifier,
      passable: passable?,
      obstructs: obstructs?
    }
  end

  def passable?
    @passable
  end

  def obstructs?
    @obstructs
  end
end